app/helpers/brexit_checker_helper.rb

```diff
@@ -1,25 +1,94 @@
 require "addressable/uri"
 
 module BrexitCheckerHelper
-  def format_criteria_list(criteria)
-    criteria.map { |criterion| { readable_text: criterion.text } }
+  def extract_criteria(object)
+    case object
+    when Array
+      object.flat_map { |element| extract_criteria(element) }
+    when Hash
+      extract_criteria(object.fetch("any_of", [])) + extract_criteria(object.fetch("all_of", []))
+    when String
+      object
+    end
   end
 
-  def format_action_audiences(actions)
+  def select_criteria(criteria, actions)
+    criteria.select do |criterion|
+      actions.any? do |action|
+        extract_criteria(action.criteria).include? criterion.key
+      end
+    end
+  end
+
+  def format_criteria(criteria, actions = [])
+    selected_criteria = if !actions.empty?
+                          select_criteria(criteria, actions)
+                        else
+                          criteria
+                        end
+    selected_criteria.map { |criterion| { readable_text: criterion.text } }
+  end
+
+  def format_action_audiences(actions, criteria)
     business, citizen = actions.partition { |action| action.audience == "business" }
-    business_results = {
-      heading: I18n.t("brexit_checker.results.audiences.business.heading"),
-      actions: order_actions_by_priority(business),
-    }
-    citizen_results = {
-      heading: I18n.t("brexit_checker.results.audiences.citizen.heading"),
-      actions: order_actions_by_priority(citizen),
-    }
-    [business_results, citizen_results]
-  end
-
-  def order_actions_by_priority(action_group)
-    action_group.sort_by.with_index do |action, index|
+    business_groups = format_business_group(business, criteria)
+    citizens_groups = format_citizen_groups(citizen, criteria)
+    business_results = if !business_groups.empty?
+                         {
+                           heading: I18n.t("brexit_checker.results.audiences.business.heading"),
+                           groups: business_groups,
+                         }
+                       end
+    citizen_results = if !citizens_groups.empty?
+                        {
+                          heading: I18n.t("brexit_checker.results.audiences.citizen.heading"),
+                          groups: citizens_groups,
+                        }
+                      end
+    [business_results, citizen_results].compact
+  end
+
+  def format_business_group(actions, criteria)
+    business_actions = order_actions_by_priority(actions)
+    if business_actions.any?
+      [{
+        heading: nil,
+        priority: 10,
+        actions: business_actions,
+        criteria: format_criteria(criteria, business_actions),
+      }]
+    else
+      []
+    end
+  end
+
+  def format_citizen_groups(actions, criteria)
+    all_groups = actions.map(&:groups).flatten.uniq
+    citizen_groups = all_groups.map do |group|
```

Comment:
I would be tempted to use ```group_by``` on actions here. That converts it into a hash with the group name as key and lists of actions as keys. That would avoid the repetitive select in ```format_citizen_actions```

Reply:
Lets run through all this grouping flow?